File: twoPointers/leetcode_1679/solution.py
from collections import defaultdict
class Solution:
    def maxOperations(self, nums: list[int], k: int) -> int:
        self.operation_count = 0
        # Solution 1 with more space but quicker
        remainings = defaultdict(int)
        nums_pointer = 0
        
        while nums_pointer < len(nums):
            # print(f"num idx: {nums_pointer}")
            # print(remainings)
            num = nums[nums_pointer]
            if remainings[num] == 0: # cannot find
                remainings[k - num] += 1
            else:
                remainings[num] -= 1
                self.operation_count += 1
            nums_pointer += 1
        return self.operation_count
    
        # Solution 2: sort the array and use two pointer

File: twoPointers/leetcode_1679/solution.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def max_operations(nums, k)
    nums.sort!

    l, r = 0, nums.length - 1
    operation = 0
    while l < r
        sum = nums[l] + nums[r]
        if sum == k
            operation += 1
            l += 1
            r -= 1
        elsif sum < k
            l += 1
        else
            r -= 1
        end
    end

    return operation
end


test_cases = [
    {
        nums: [1,2,3,4],
        k: 5
    },
    {
        nums: [3,1,3,4,3],
        k: 6
    }
]

for test_case in test_cases
    print(max_operations(test_case[:nums], test_case[:k]))
end
